File: c++/Longest ZigZag Path in a Binary Tree.cpp
#include <iostream>
#include <algorithm>

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    int max_length;

    Solution() : max_length(0) {}

    void dfs(TreeNode* node, bool isLeft, int length) {
        if (!node) return;

        max_length = std::max(max_length, length);

        if (isLeft) {
            dfs(node->left, false, length + 1);
            dfs(node->right, true, 1);        
        } else {
            dfs(node->right, true, length + 1); 
            dfs(node->left, false, 1);          
        }
    }

    int longestZigZag(TreeNode* root) {
        dfs(root, true, 0);  
        dfs(root, false, 0); 
        return max_length;
    }
};
//  test case defs
void test_longestZigZag() {
    TreeNode* root = new TreeNode(1);
    root->left = new TreeNode(2);
    root->right = new TreeNode(3);
    root->left->right = new TreeNode(4);
    root->left->right->left = new TreeNode(6);
    root->right->right = new TreeNode(5);
    root->right->right->right = new TreeNode(7);

    Solution solution;
    int result = solution.longestZigZag(root);

    std::cout << "Test case result: " << result << std::endl;

    delete root->right->right->right;
    delete root->right->right;
    delete root->right;
    delete root->left->right->left;
    delete root->left->right;
    delete root->left;
    delete root;
}

int main() {
    test_longestZigZag();
    return 0;
}

File: python/Longest ZigZag Path in a Binary Tree.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def __init__(self):
        self.max_length = 0

    def longestZigZag(self, root: TreeNode) -> int:
        def dfs(node: TreeNode, direction: str, length: int):
            if not node:
                return

            self.max_length = max(self.max_length, length)

            if direction == 'left':
                dfs(node.left, 'right', length + 1)  
                dfs(node.right, 'left', 1)           
            else:
                dfs(node.right, 'left', length + 1)  
                dfs(node.left, 'right', 1)          

        dfs(root, 'left', 0)
        dfs(root, 'right', 0)
        
        return self.max_length

# Test case
def test_longestZigZag():
    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(3)
    root.left.right = TreeNode(4)
    root.left.right.left = TreeNode(6)
    root.right.right = TreeNode(5)
    root.right.right.right = TreeNode(7)

    solution = Solution()

    result = solution.longestZigZag(root)

    print(f"Test case result: {result}")

test_longestZigZag()
